typing/typemod.ml

```diff
@@ -85,33 +87,112 @@ let extract_sig_open env loc mty =
 
 (* Compute the environment after opening a module *)
 
-let type_open_ ?used_slot ?toplevel ovf env loc lid =
-  let path = Typetexp.lookup_module ~load:true env lid.loc lid.txt in
-  match Env.open_signature ~loc ?used_slot ?toplevel ovf path env with
-  | Some env -> path, env
-  | None ->
-      let md = Env.find_module path env in
-      ignore (extract_sig_open env lid.loc md.md_type);
-      assert false
+let type_module_fwd : (Env.t -> Parsetree.module_expr ->
+                       Typedtree.module_expr) ref =
+  ref (fun _ _ -> assert false)
+
+let mod_ident_counter = ref 0
+let generated_module_ident = ref []
+
+let push_current_mid (mi, md, env) =
+  generated_module_ident := (mi, md, env) :: !generated_module_ident
+
+let pop_current_mid () =
+  let slot = List.hd !generated_module_ident in
+  generated_module_ident := List.tl !generated_module_ident;
+  slot
+
+let gen_mod_ident () =
+  let n = !mod_ident_counter in
+  incr mod_ident_counter;
+  let ident = Ident.create (Printf.sprintf "M#%d" n) in
+  ident
+
+let open_struct_level = ref 0
+let in_nested_struct () = !open_struct_level <> 0
+let enter_struct () = incr open_struct_level
+let leave_struct () = decr open_struct_level
+
+let type_open_ ?used_slot ?toplevel ovf env loc me =
+  match me.pmod_desc with
+  | Pmod_functor _ | Pmod_unpack _ | Pmod_extension _ ->
+      raise(Error(me.pmod_loc, env, Invalid_open me))
+  | Pmod_ident lid -> begin
+      let path = Typetexp.lookup_module ~load:true env lid.loc lid.txt in
+      match Env.open_signature ~loc ?used_slot ?toplevel ovf path env with
+      | Some env ->
+          let tme =
+            {
+              mod_desc=Tmod_ident (path, lid);
+              mod_loc=lid.loc;
+              mod_type=Mty_ident path;
+              mod_env=env;
+              mod_attributes=me.pmod_attributes
+            } in
+          tme, env
+      | None ->
+          let md = Env.find_module path env in
+          ignore (extract_sig_open env lid.loc md.md_type);
+          assert false
```

Comment:
I know the code was here originally, but it may be nice to add a comment explaining that `extract_sig_open` raises an `structure expected` since the opened module is a functior in this branch.

Reply:
I have refactored so that extraction is no longer needed